File: Backend/api/dataRoutes.js
const express = require('express');
const DataController = require('../controllers/dataController');
const auth = require('../middlewares/auth');
const role = require('../middlewares/role');

const router = express.Router();
router.use(auth);

router.get('/', DataController.getAll);
router.get('/latest-per-lokasi', DataController.getLatestPerLokasi);
router.get('/latest', DataController.getLatest);
router.get('/lastday', DataController.getLastDay);
router.get('/stats', DataController.getStats);
router.get('/:id', DataController.getById);

router.post('/', role(['admin', 'user']), DataController.create);
router.put('/:id', role(['admin', 'user']), DataController.update);
router.delete('/:id', role(['admin', 'user']), DataController.delete);

module.exports = router;

File: backend_python/api/data_routes.py
from flask import Blueprint

from controllers.data_controller import DataController
from middlewares.auth import auth_required, role_required

bp = Blueprint("data", __name__)

GET_ROUTES = [
    ("/", "get_all", DataController.get_all),
    ("/latest-per-lokasi", "latest_per_lokasi", DataController.get_latest_per_lokasi),
    ("/latest", "latest", DataController.get_latest),
    ("/lastday", "lastday", DataController.get_last_day),
    ("/stats", "stats", DataController.get_stats),
]

for path, endpoint, view in GET_ROUTES:
    bp.add_url_rule(path, endpoint=endpoint, view_func=auth_required(view), methods=["GET"])

WRITE_ROLES = role_required(["admin", "user"])

bp.add_url_rule(
    "/<int:row_id>", endpoint="get_by_id", view_func=auth_required(DataController.get_by_id), methods=["GET"]
)
bp.add_url_rule(
    "/", endpoint="create", view_func=auth_required(WRITE_ROLES(DataController.create)), methods=["POST"]
)
bp.add_url_rule(
    "/<int:row_id>", endpoint="update", view_func=auth_required(WRITE_ROLES(DataController.update)), methods=["PUT"]
)
bp.add_url_rule(
    "/<int:row_id>",
    endpoint="delete",
    view_func=auth_required(WRITE_ROLES(DataController.delete_row)),
    methods=["DELETE"],
)
